### PR TITLE
Segmented move implemented for hpx::vector

### PR DESCRIPTION
**Pull request by Sanac**
This should resolve "move" bullet point for #1338.

- Segmented move implemented based on changes made in `parallel::copy`.
- Refactored so that both algorithms (`parallel::copy` and `parallel::move`) are based on the same code.
- Changed return values of `parallel::move` to match `parallel::copy`

### CODE REVIEW

**Comment by hkaiser** [comment]
Wow! Many thanks. I will look closer over the next days.

**Comment by sithhell** [comment]
There is a problem with `hpx::lcos::local::dataflow`, please change it to fully qualify for `hpx::dataflow`

**Comment by Sanac** [comment]
Compile error should be fixed now. Thanks for your help sithhell!

File: hpx/include/parallel_copy.hpp [hunk 1]
```diff
@@ -8,8 +8,7 @@
 #define HPX_PARALLEL_COPY_JUN_28_2014_0827AM
 
 #include <hpx/parallel/algorithms/copy.hpp>
-#include <hpx/parallel/segmented_algorithms/copy.hpp>
-#include <hpx/parallel/container_algorithms/copy.hpp>
+#include <hpx/parallel/container_algorithms.hpp>
```
Comment (hkaiser) [inline hpx/include/parallel_copy.hpp]: What is the rationale of this change? Why include all container algorithms in this header file?

File: hpx/parallel/algorithms/copy.hpp [hunk 2]
```diff
@@ -13,6 +13,7 @@
 #include <hpx/traits/concepts.hpp>
 #include <hpx/util/invoke.hpp>
 #include <hpx/util/tagged_pair.hpp>
+#include <hpx/traits/segmented_iterator_traits.hpp>
```
Comment (hkaiser) [inline hpx/parallel/algorithms/copy.hpp]: This header is not needed anymore in this file.

File: hpx/parallel/algorithms/copy.hpp [hunk 3]
```diff
@@ -26,6 +27,8 @@
 #include <hpx/parallel/util/projection_identity.hpp>
 #include <hpx/parallel/util/zip_iterator.hpp>
 #include <hpx/parallel/traits/projected.hpp>
+#include <hpx/parallel/algorithms/transfer.hpp>
+#include <hpx/traits.hpp>
```
Comment (hkaiser) [inline hpx/parallel/algorithms/copy.hpp]: Why have you included this header file? It forward declares all traits only. If you need a specific trait, include the corresponding header instead.

File: hpx/parallel/algorithms/copy.hpp [hunk 4]
```diff
@@ -187,34 +158,12 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
...
-        typedef typename iterator_traits::is_segmented_iterator is_segmented;
-
-        return hpx::util::make_tagged_pair<tag::in, tag::out>(
-            detail::copy_(
-                std::forward<ExPolicy>(policy), first, last, dest,
-                is_segmented()));
+            return transfer<
+                detail::copy
+                >(std::forward<ExPolicy>(policy), first, last, dest);
     }
```
Comment (hkaiser) [inline hpx/parallel/algorithms/copy.hpp]: Please fix indentation here.

File: hpx/parallel/algorithms/move.hpp [hunk 5]
```diff
@@ -119,41 +133,18 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
...
-            boost::is_same<std::input_iterator_tag, input_iterator_category>,
-            boost::is_same<std::output_iterator_tag, output_iterator_category>
-        >::type is_seq;
-
-        return detail::move<OutIter>().call(
-            std::forward<ExPolicy>(policy), is_seq(),
-            first, last, dest);
+            return transfer<detail::move>(
+                std::forward<ExPolicy>(policy), first, last, dest);
     }
```
Comment (hkaiser) [inline hpx/parallel/algorithms/move.hpp]: Please fix indentation here.

File: hpx/parallel/algorithms/transfer.hpp [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        typedef typename iterator_traits::is_segmented_iterator is_segmented;
+
+        return hpx::util::make_tagged_pair<tag::in, tag::out>(
+            detail::transfer_<Algo>(
+                std::forward<ExPolicy>(policy), first, last, dest,
+                is_segmented())
+            );
+    }
+}}}
+
```
Comment (hkaiser) [inline hpx/parallel/algorithms/transfer.hpp]: I'd suggest to move the transfer() algorithm into the detail namespace as it is not  to be exposed to users.

File: hpx/parallel/algorithms/transfer.hpp [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    ///           \a sequential_task_execution_policy or
+    ///           \a parallel_task_execution_policy and
+    ///           returns \a OutIter otherwise.
+    ///           The \a move algorithm returns the output iterator to the
+    ///           element in the destination range, one past the last element
+    ///           transfered.
+    ///
+    template <
+        template <typename IterPair> class Algo,
+        typename ExPolicy, typename InIter, typename OutIter,
```
Comment (hkaiser) [inline hpx/parallel/algorithms/transfer.hpp]: I'd suggest not to use a template template argument here. I think you could pass the fully specified type of the algorithm implementation instead.

File: hpx/parallel/algorithms/move.hpp [hunk 8]
```diff
@@ -119,41 +133,18 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
     ///           element in the destination range, one past the last element
     ///           copied.
     ///
```
Comment (hkaiser) [inline hpx/parallel/algorithms/move.hpp]: Please adapt the documentation to reflect the change in return type for the `parallel::move()` algorithm.

File: hpx/parallel/algorithms/transfer.hpp [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    /// \returns  The \a transfer algorithm returns a \a hpx::future<OutIter> if
+    ///           the execution policy is of type
+    ///           \a sequential_task_execution_policy or
+    ///           \a parallel_task_execution_policy and
+    ///           returns \a OutIter otherwise.
+    ///           The \a move algorithm returns the output iterator to the
+    ///           element in the destination range, one past the last element
+    ///           transfered.
+    ///
+    template <
```
Comment (hkaiser) [inline hpx/parallel/algorithms/transfer.hpp]: Please adapt this documentation to reflect the correct return type. Also, this comment should not be used for doxygen documentation generation, please change `///` to `//`.

File: hpx/parallel/algorithms/transfer.hpp (hunk 7 above)
Comment (hkaiser) [inline hpx/parallel/algorithms/transfer.hpp]: Actually, the whole header could go into `parallel/algorithms/detail`.

**Comment by Sanac** [comment]
- Clean up of: unneeded includes, indentation, documentation
- Moved `transfer.hpp`'s into `algorithms/detail` and `segmented_algorithms/detail`
- Use of fully declared algorithm instead of using template template argument

**Comment by hkaiser** [comment]
Very nice! LGTM, thanks!

**Comment by Sanac** [comment]
Merged.
Thanks for your help @hkaiser !
